spec/java/util/zip/CRC32.automaton.lsl

```diff
@@ -0,0 +1,202 @@
+import sun/nio/ch/DirectBuffer;
+import java/util/zip/CRC32;
+
+// automata
+
+automaton CRC32Automaton
+(
+)
+: LSLCRC32
+{
+    // states and shifts
+
+    initstate Allocated;
+    state Initialized;
+
+    shift Allocated -> Initialized by [
+        // constructors
+        CRC32,
+    ];
+
+    shift Initialized -> self by [
+        // instance methods
+        getValue,
+        reset,
+        update (CRC32, ByteBuffer),
+        update (CRC32, array<byte>),
+        update (CRC32, array<byte>, int, int),
+        update (CRC32, int),
+    ];
+
+    // internal variables
+
+    var crc: int = 0;
+
+
+    // utilities
+
+    proc _updateCheck(b: array<byte>, off: int, len: int): void
+    {
+        if (b == null)
+            action THROW_NEW("java.lang.NullPointerException", []);
+
+        val b_size: int = action ARRAY_SIZE(b);
+        if (off < 0 || len < 0 || off > b_size - len)
+            action THROW_NEW("java.lang.ArrayIndexOutOfBoundsException", []);
+    }
+
+
+    proc _updateByteBuffer(addr: long): int
+    {
+        _updateByteBufferCheck(addr);
+        result = action SYMBOLIC("int");
+    }
+
+
+    @AutoInline @Phantom proc _updateByteBufferCheck(addr: long): void
+    {
+        if (addr == 0L)
+            action THROW_NEW("java.lang.NullPointerException", []);
+    }
+
+
+    proc _updateBytes(b: array<byte>, off: int, len: int): int
+    {
+        _updateBytesCheck(b, off, len);
+        result = action SYMBOLIC("int");
+    }
+
+
+    @AutoInline @Phantom proc _updateBytesCheck (b: array<byte>, off: int, len: int): void
+    {
+        if (len != 0)
+        {
+            if (b != null)
+            {
+                val b_size: int = action ARRAY_SIZE(b);
+                if (off < 0 || off >= b_size)
+                    action THROW_NEW("java.lang.ArrayIndexOutOfBoundsException", [off]);
+
+                val endIndex: int = off + len -1;
+                if (endIndex <0 || endIndex >= b_size)
+                    action THROW_NEW("java.lang.ArrayIndexOutOfBoundsException", [endIndex]);
+            }
+            else
+            {
+                action THROW_NEW("java.lang.NullPointerException", []);
+            }
+        }
+    }
+
+    // constructors
+
+    constructor *.CRC32 (@target self: CRC32)
+    {
+        // original constructor is empty
+    }
+
+
+    // static methods
+
+    // methods
+
+    fun *.getValue (@target self: CRC32): long
+    {
+        result = (this.crc as long) & 4294967295L;
+    }
+
+
+    fun *.reset (@target self: CRC32): void
+    {
+        this.crc = 0;
+    }
+
+
+    fun *.update (@target self: CRC32, buffer: ByteBuffer): void
+    {
+        val pos: int = action CALL_METHOD(buffer, "position", []);
+        val limit: int = action CALL_METHOD(buffer, "limit", []);
+        if (pos > limit)
+            action THROW_NEW("java.lang.AssertionError", []);
+        // #warning: assert (pos <= limit) can be disabled in original CRC32
```

Comment:
`...but this cannot`

I also suggest placing this comment on top of `action THROW_NEW...` instead of under it.